tests/Unit/Repositories/CampaignTenantRepositoryTest.php

```diff
@@ -80,11 +95,83 @@ public function the_average_time_to_click_attribute_returns_na_if_there_have_bee
 
         static::assertEquals('N/A', $averageTimeToClick);
     }
 
+    /** @test */
+    public function the_cancel_campaign_method_sets_the_campaign_status_to_cancelled()
+    {
+        $campaign = factory(Campaign::class)->state('queued')->create();
+
+        static::assertEquals(CampaignStatus::STATUS_QUEUED, $campaign->status_id);
+        $success = app(CampaignTenantRepositoryInterface::class)->cancelCampaign($campaign);
+
+        static::assertTrue($success);
+        static::assertEquals(CampaignStatus::STATUS_CANCELLED, $campaign->fresh()->status_id);
+    }
+
+    /** @test */
+    public function the_cancel_campaign_method_deletes_draft_messages_if_the_campaign_has_any()
+    {
+        [$workspace, $emailService] = $this->createUserWithWorkspaceAndEmailService();
+
+        $campaign = factory(Campaign::class)->states(['withContent', 'sent'])->create([
+            'workspace_id' => $workspace->id,
+            'email_service_id' => $emailService->id,
+            'save_as_draft' => 1,
+        ]);
+        $this->createPendingMessage($workspace, $campaign, 1);
+
+        static::assertCount(1, Message::all());
+
+        $this->campaignRepository->cancelCampaign($campaign);
+
+        static::assertCount(0, Message::all());
+    }
+
+    /** @test */
+    public function the_cancel_campaign_method_does_not_delete_sent_messages()
+    {
+        [$workspace, $emailService] = $this->createUserWithWorkspaceAndEmailService();
+
+        $campaign = factory(Campaign::class)->states(['withContent', 'sent'])->create([
+            'workspace_id' => $workspace->id,
+            'email_service_id' => $emailService->id,
+            'save_as_draft' => 1,
+        ]);
+        $this->createOpenedMessage($workspace, $campaign, 1);
+
+        static::assertCount(1, Message::all());
+
+        $this->campaignRepository->cancelCampaign($campaign);
+
+        static::assertCount(1, Message::all());
+    }
+
+    /** @test */
+    public function the_get_count_method_returns_campaign_message_counts()
+    {
+        [$workspace, $emailService] = $this->createUserWithWorkspaceAndEmailService();
+        $campaign = $this->createCampaign($workspace, $emailService);
+
+        $openedMessages = $this->createOpenedMessage($workspace, $campaign, 1);
+        $sentMessages = $this->createUnopenedMessage($workspace, $campaign, 2);
+        $clickedMessages = $this->createClickedMessage($workspace, $campaign, 3);
+        $bouncedMessages = $this->createBouncedMessage($workspace, $campaign, 4);
+        $pendingMessages = $this->createPendingMessage($workspace, $campaign, 5);
+
+        $counts = $this->campaignRepository->getCounts(collect($campaign->id), $workspace->id);
+
+        $totalSentCount = $openedMessages->count() + $clickedMessages->count() + $sentMessages->count() + $bouncedMessages->count();
+        static::assertEquals($openedMessages->count(), $counts[$campaign->id]->opened);
```

Comment:
You can be explicit here and set the number instead of retrieving it from the collection. This allows you also to catch errors in the `createOpenedMessge` implementation. Also I would rename all those methods with the plural form `*Messages`